### PR TITLE
unpack dynamic kubelet config payloads to files

### PR DESCRIPTION
**Pull request by mtaufen**
This PR unpacks the downloaded ConfigMap to a set of files on the node.

This enables other config files to ride alongside the
KubeletConfiguration, and the KubeletConfiguration to refer to these
cohabitants with relative paths.

This PR also stops storing dynamic config metadata (e.g. current,
last-known-good config records) in the same directory as config
checkpoints. Instead, it splits the storage into `meta` and
`checkpoints` dirs.

The current store dir structure is as follows:
```
- dir named by --dynamic-config-dir (root for managing dynamic config)
| - meta (dir for metadata, e.g. which config source is currently assigned, last-known-good)
  | - current (a serialized v1 NodeConfigSource object, indicating the assigned config)
  | - last-known-good (a serialized v1 NodeConfigSource object, indicating the last-known-good config)
| - checkpoints (dir for config checkpoints)
  | - uid1 (dir for unpacked config, identified by uid1)
    | - file1
    | - file2
    | - ...
  | - uid2
  | - ...
```

There are some likely changes to the above structure before dynamic config goes beta, such as renaming "current" to "assigned" for clarity, and extending the checkpoint identifier to include a resource version, as part of resolving #61643.

```release-note
NONE
```

/cc @luxas @smarterclayton

### CODE REVIEW

**Comment by smarterclayton** [comment]
@liggitt as well.

File: pkg/kubelet/kubeletconfig/util/files/files_test.go [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+
+	utiltest "k8s.io/kubernetes/pkg/kubelet/kubeletconfig/util/test"
+	utilfs "k8s.io/kubernetes/pkg/util/filesystem"
+)
+
+const (
+	prefix = "test-util-files"
+)
+
+type file struct {
```
Comment (mtaufen) [inline pkg/kubelet/kubeletconfig/util/files/files_test.go]: @tianshapjq this file is what I'm thinking for the files tests, let me know what you think

**Comment by BenTheElder** [comment]
[caching experiment, please ignore]
/test pull-kubernetes-bazel-test-canary
/test pull-kubernetes-bazel-build-canary

**Comment by jberkus** [comment]
can we move this issue towards approval for Code Freeze on Feb 26th?  

@dchen1107 @mtaufen

**Comment by jberkus** [comment]
Reminder: this PR needs status/approved-for-milestone, or it will be automatically kicked on at Code Freeze (monday).

File: pkg/kubelet/kubeletconfig/util/files/files_test.go (hunk 1 above)
Comment (tianshapjq) [reply pkg/kubelet/kubeletconfig/util/files/files_test.go]: TBH, why do you try to test the Create(), Write() and MkdirAll() of the DefaultFs again? it's already been done in the files.go, and we just have to write some unit tests to the functions in files.go IMH.
BTW, you may need to remove the files created by this test case.

Comment (mtaufen) [reply pkg/kubelet/kubeletconfig/util/files/files_test.go]: > TBH, why do you try to test the Create(), Write() and MkdirAll() of the DefaultFs again? it's already been done in the files.go, and we just have to write some unit tests to the functions in files.go IMH.

I'm not sure I follow? This file is designed to unit test the functions in files.go.

> BTW, you may need to remove the files created by this test case.

Agree, will add some cleanup.

Comment (tianshapjq) [reply pkg/kubelet/kubeletconfig/util/files/files_test.go]: sorry I didn't make it clear. My point is, you have tested the Create(), Write() and MkdirAll() here, seems my test case #57220 is useless, because my test case just invokes these functions as well. So, is my test case still needed?

Comment (mtaufen) [reply pkg/kubelet/kubeletconfig/util/files/files_test.go]: I understood #57220 as an effort to test the other utility functions, not the filesystem functions (`FileExists`, `EnsureFile`, `WriteTmpFile`, `ReplaceFile`, `DirExists`, `EnsureDir`).

**Comment by BenTheElder** [comment]
@mtaufen has a typo and needs a hack/update-bazel.sh

File: pkg/kubelet/kubeletconfig/util/files/files_test.go (hunk 1 above)
Comment (tianshapjq) [reply pkg/kubelet/kubeletconfig/util/files/files_test.go]: ok, I will update that pr after this merged~

File: pkg/kubelet/kubeletconfig/util/codec/codec.go [hunk 2]
```diff
@@ -26,9 +26,37 @@ import (
...
+	// encoder will convert to external version
+	data, err := runtime.Encode(encoder, internal)
+	if err != nil {
+		return nil, err
+	}
+	return data, nil
+}
+
+// NewKubeletConfigJSONEncoder returns an encoder that can write a KubeletConfig to JSON
+func NewKubeletConfigJSONEncoder() (runtime.Encoder, error) {
```
Comment (dashpole) [inline pkg/kubelet/kubeletconfig/util/codec/codec.go]: I would rather not export this if it is only used by `EncodeKubeletConfig`.  We can always make it an exported method at a later date.

File: pkg/kubelet/kubeletconfig/util/files/files.go [hunk 3]
```diff
@@ -24,7 +24,10 @@ import (
 	utilfs "k8s.io/kubernetes/pkg/util/filesystem"
 )
 
-const defaultPerm = 0666
+const (
+	defaultPerm = 0755
```
Comment (dashpole) [inline pkg/kubelet/kubeletconfig/util/files/files.go]: can you add a brief comment explaining the permissions?  Why do we want it to be executable, and what is it the permission for?

**Review by dashpole — COMMENTED** [review]
mtaufen and I discussed the design offline.  Just a couple nits, otherwise it seems pretty straightforward.

File: pkg/kubelet/kubeletconfig/util/files/files.go (hunk 3 above)
Comment (mtaufen) [reply pkg/kubelet/kubeletconfig/util/files/files.go]: If you don't make a directory executable, you can't access the items in it. I thought about doing things differently for files vs. dirs, but decided to keep it uniform for simplicity.

With directories, AFAIK:
- `r`: allows listing contained files (well, sort of, you'll see the file names but get other permission errors from `ls` if this is all you have, since you need `x` to read the permission bits, size, timestamps, etc. for the files).
- `w`: allows creating files (in practice you often also need `x` to effectively create files, e.g `touch` fails without `x`, I don't actually know if there's a scenario where `w` is useful on its own)
- `x`: allows accessing files (e.g. `cat`) and changing into the directory (e.g. `cd`)

File: pkg/kubelet/kubeletconfig/util/codec/codec.go (hunk 2 above)
Comment (mtaufen) [reply pkg/kubelet/kubeletconfig/util/codec/codec.go]: I think I had planned on refactoring the e2e node tests to use this, instead of the duplicate in util.go. I will add that to this PR.

Comment (mtaufen) [reply pkg/kubelet/kubeletconfig/util/codec/codec.go]: Updated.

File: pkg/kubelet/kubeletconfig/util/files/files.go (hunk 3 above)
Comment (dashpole) [reply pkg/kubelet/kubeletconfig/util/files/files.go]: good to know.  Was mostly just curious.

**Comment by dashpole** [comment]
took another look over.
/lgtm

**Comment by mtaufen** [comment]
rebased

**Comment by mtaufen** [comment]
re-added label post rebase

File: pkg/kubelet/kubeletconfig/util/codec/codec.go [hunk 4]
```diff
@@ -26,9 +26,37 @@ import (
...
+	return data, nil
+}
+
+// newKubeletConfigJSONEncoder returns an encoder that can write a KubeletConfig to JSON
+func newKubeletConfigJSONEncoder() (runtime.Encoder, error) {
+	_, codecs, err := scheme.NewSchemeAndCodecs()
+	if err != nil {
+		return nil, err
+	}
+	mediaType := "application/json"
```
Comment (liggitt) [inline pkg/kubelet/kubeletconfig/util/codec/codec.go]: I know this is a straight move, but did we want json or yaml? not a blocking issue, just curious... config files elsewhere have preferred yaml

File: pkg/kubelet/kubeletconfig/util/codec/codec.go [hunk 5]
```diff
@@ -26,9 +26,37 @@ import (
 	"k8s.io/apimachinery/pkg/runtime/serializer"
 	"k8s.io/kubernetes/pkg/api/legacyscheme"
 	"k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig"
+	"k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig/scheme"
+	"k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig/v1beta1"
 )
 
-// TODO(mtaufen): allow an encoder to be injected into checkpoint objects at creation time? (then we could ultimately instantiate only one encoder)
+// EncodeKubeletConfig encodes an internal KubeletConfiguration to an external JSON representation
+func EncodeKubeletConfig(internal *kubeletconfig.KubeletConfiguration) ([]byte, error) {
```
Comment (liggitt) [inline pkg/kubelet/kubeletconfig/util/codec/codec.go]: if this is moving to non-test code, do we need to allow specifying the target version here?

File: pkg/kubelet/kubeletconfig/checkpoint/store/fsstore.go [hunk 6]
```diff
@@ -21,82 +21,99 @@ import (
...
 }
 
-func (s *fsStore) Load(uid string) (checkpoint.Checkpoint, error) {
-	filePath := filepath.Join(s.checkpointsDir, uid)
-	utillog.Infof("loading configuration from %q", filePath)
-
-	// load the file
-	data, err := s.fs.ReadFile(filePath)
+func (s *fsStore) Load(source checkpoint.RemoteConfigSource) (*kubeletconfig.KubeletConfiguration, error) {
+	sourceFmt := fmt.Sprintf("%s:%s", source.APIPath(), source.UID())
```
Comment (liggitt) [inline pkg/kubelet/kubeletconfig/checkpoint/store/fsstore.go]: is `:` a safe filesystem character? on windows as well?

File: pkg/kubelet/kubeletconfig/util/codec/codec.go (hunk 4 above)
Comment (mtaufen) [reply pkg/kubelet/kubeletconfig/util/codec/codec.go]: Agree we should probably update this to use yaml.

File: pkg/kubelet/kubeletconfig/util/codec/codec.go (hunk 5 above)
Comment (mtaufen) [reply pkg/kubelet/kubeletconfig/util/codec/codec.go]: It's not necessary yet. Maybe in the future, when multiple versions are available.

File: pkg/kubelet/kubeletconfig/checkpoint/store/fsstore.go (hunk 6 above)
Comment (mtaufen) [reply pkg/kubelet/kubeletconfig/checkpoint/store/fsstore.go]: Ah, darn, not safe on Windows. I'll update. Thanks.

Comment (mtaufen) [reply pkg/kubelet/kubeletconfig/checkpoint/store/fsstore.go]: Oh, hold on, this is just for logging. Not written to fs. Too long since I last read this code xD.

Comment (liggitt) [reply pkg/kubelet/kubeletconfig/checkpoint/store/fsstore.go]: ah, misread

File: pkg/kubelet/kubeletconfig/util/codec/codec.go (hunk 5 above)
Comment (liggitt) [reply pkg/kubelet/kubeletconfig/util/codec/codec.go]: > Maybe in the future, when multiple versions are available.

I don't want callers to silently get the version they persist changed out from under them... making the version written implicit leads to that.

Comment (mtaufen) [reply pkg/kubelet/kubeletconfig/util/codec/codec.go]: Fair enough, updated.

**Comment by liggitt** [comment]
/approve

didn't review the checkpoint dir changes closely, will defer to @dashpole's review on that

can you summarize in the description the structure of the store dir structure and file contents? before moving to beta, we'll want to hoist that documentation out to official docs, since we'll need to remain compatible across release boundaries

**Comment by BenTheElder** [comment]
xref: https://github.com/kubernetes/test-infra/pull/7791

**Comment by mtaufen** [comment]
@liggitt will do, thanks

**Comment by dashpole** [comment]
took another pass
/lgtm

**Comment by mtaufen** [comment]
pull-kubernetes-integration is fixed by https://github.com/kubernetes/kubernetes/pull/63017, will retest once that goes in
